src/lmdb.ml

```diff
@@ -162,21 +170,81 @@ module Env = struct
 
 end
 
-(* Use internally for trivial functions *)
-let trivial_txn ~write env f =
-  let txn = alloc mdb_txn in
-  let txn_flag = if write
-    then Env.Flags.none
-    else Env.Flags.read_only
-  in
-  mdb_txn_begin env None txn_flag txn ;
-  try
-    let x = f !@txn in
-    mdb_txn_commit !@txn ;
-    x
-  with e ->
-    mdb_txn_abort !@txn ;
-    raise e
+module Txn :
+sig
+  type -'a t = mdb_txn constraint 'a = [< `Read | `Write ]
+
+  val go :
+    ?parent:([< `Read | `Write ] as 'a) t ->
+    'a cap ->
+    Env.t ->
+    ('a t -> 'b) -> 'b option
+
+  val abort : 'a t -> 'b
+
+  val env : 'a t -> Env.t
+
+  (* not exported: *)
+  val maybe :
+    'a cap ->
+    ?txn:'a t ->
+    Env.t ->
+    ('a t -> 'b) -> 'b
+end
+=
+struct
+  type -'a t = mdb_txn constraint 'a = [< `Read | `Write ]
+
+  exception Abort of mdb_txn
+
+  let env txn = mdb_txn_env txn
+
+  let abort txn = raise (Abort txn)
+
+  let go :
+    ?parent: 'a t ->
+    'a cap ->
+    Env.t ->
+    ('a t -> 'b) -> 'b option
+    =
+    fun ?parent rw env f ->
+    let ptr_txn = alloc mdb_txn in
+    let txn_flag =
+      rw |> fun (type a) (rw :a cap) -> match rw with
```

Comment:
These type annotations are both unnecessary and rather non idiomatic. Just move the `(type a)` on the outer function and simplify this.

Reply:
done.